src/dataflow-types/src/types.rs

```diff
@@ -762,6 +762,28 @@ pub mod sources {
         User(String),
     }
 
+    /// `SourceEnvelope`'s, describe how to turn a stream of messages from `SourceConnector`'s,
+    /// and turn them into a _differential stream_, that is, a stream of (data, time, diff)
+    /// triples.
+    ///
+    /// The most trivial version is `None`, which typically produces triples where the diff
+    /// is ALWAYS `1`
+    /// TODO(guswynn): update `None` docs to describe `-1` diff cases
+    ///
+    /// `Upsert` holds onto previously seen values and produces `1` or `-1` diffs depending on
+    /// whether or not the required _key_ outputed by the source has been seen before. This also
+    /// supports a `Debezium` mode.
+    ///
+    /// `Debezium` avoids holding onto previously seen values by trusting the required
+    /// `before` and `after` value fields coming from the upstream source.
+    ///
+    /// `CdcV2` requires sources output messages in a strict form that requires a upstream-provided
+    /// timeline.
+    ///
+    /// Some sources (namely postgres and pubnub) skip any explicit envelope handling, effectively
+    /// asserting that `SourceEnvelope` is `None` with `KeyEnvelope::None`. (TODO(guswynn): improve
+    /// the complex way we ensure this in `sql::plan::statement::ddl`)
+    /// TODO(guswynn): update this ^ when SimpleSource is gone.
```

Comment:
This last paragraph is the kind of thing that is almost certain to go stale, because no one will think to update the `SourceEnvelope` docstring when they fix `sql::plan::statement::ddl`. Can we move this TODO to `sql::plan::statement::ddl` instead? The comment I'd leave here is something along the lines of:

> Not all source envelopes are compatible with all source connectors. Whoever constructs the source ingestion pipeline is responsible for choosing compatible envelopes and connectors.